File: src/services/scheduler.ts
/**
 * Scheduler: fires scheduled/recurring tasks and named cron automations
 * (§15). Also releases follow-up chain steps. A `Clock` abstraction lets
 * tests advance time and force ticks deterministically.
 *
 * Cron format: `minute hour day-of-month month day-of-week` (5 fields).
 * Supports `*`, numbers, lists, ranges and step values (e.g. `*`/`5`)
 * (enough for the blueprint's examples: weekly, first-of-month).
 */
import { eq, and } from "drizzle-orm";
import { tasks, automations, agents, taskTemplates } from "../db/schema.js";
import type { DB } from "../db/client.js";
import type { Services } from "./registry.js";
import type { Task } from "../domain/types.js";

export interface Clock {
  now(): Date;
}

export class RealClock implements Clock {
  now(): Date {
    return new Date();
  }
}

export class SchedulerService {
  private timer: ReturnType<typeof setInterval> | null = null;
  private ticking = false;
  private cloudBusy = false;
  private dueMarked = new Set<string>(); // taskIds already dispatched (until status changes)

  constructor(
    private db: DB,
    private services: Services,
    private clock: Clock = new RealClock(),
    private intervalMs = 1000,
  ) {}

  start(): void {
    if (this.timer) return;
    this.timer = setInterval(() => {
      void this.tick().catch((e) => {
        void this.services.activity.emit({
          projectId: null,
          type: "system",
          actor: "scheduler",
          message: `tick error: ${e instanceof Error ? e.message : e}`,
        });
      });
    }, this.intervalMs);
  }

  stop(): void {
    if (this.timer) clearInterval(this.timer);
    this.timer = null;
  }

  isCloudBusy(): boolean {
    return this.cloudBusy;
  }

  setCloudBusy(busy: boolean): void {
    this.cloudBusy = busy;
  }

  /** Force a scan (tests use this with a fake clock). */
  async tick(): Promise<void> {
    if (this.ticking) return;
    this.ticking = true;
    try {
      await this.dispatchDueTasks();
      await this.fireAutomations();
    } finally {
      this.ticking = false;
    }
  }

  private async dispatchDueTasks(): Promise<void> {
    const all = await this.db.select().from(tasks).all();
    const now = this.clock.now();
    for (const row of all) {
      const task: Task = { ...row, activity: row.activity ?? [], attachmentIds: row.attachmentIds ?? [], sessionIds: row.sessionIds ?? [] };
      // Recurring cron tasks: after an agent finishes one run, reset to todo
      // so the next cron match fires again (§9.2 recurring). Falls through to
      // the dispatch check below so a match on this same tick still fires.
      if (task.status === "done" && task.scheduleKind === "cron") {
        await this.db
          .update(tasks)
          .set({
            status: "todo",
            activity: [
              ...task.activity,
              { at: now.toISOString(), actor: "scheduler", message: "recurring task reset for next run" },
            ],
          })
          .where(eq(tasks.id, task.id))
          .run();
        task.status = "todo";
      }
      if (task.status !== "todo") {
        this.dueMarked.delete(task.id);
        continue;
      }
      if (task.assigneeType !== "agent" || !task.assigneeAgentId) continue;
      if (this.dueMarked.has(task.id)) continue;
      if (this.services.sessions.isTaskRunning(task.id)) continue;
      if (task.chainId && task.chainIndex !== null && task.chainIndex > 0) {
        // Chain: blocked until the previous step is done (§9.4).
        const prev = all
          .filter((t) => t.chainId === task.chainId && t.chainIndex === task.chainIndex! - 1)
          .sort((a, b) => (a.chainIndex ?? 0) - (b.chainIndex ?? 0))[0];
        if (!prev || prev.status !== "done") continue;
      }
      let due = false;
      if (task.scheduleKind === "now") due = true;
      else if (task.scheduleKind === "at" && task.runAt) due = now >= new Date(task.runAt);
      else if (task.scheduleKind === "cron" && task.cron && task.timezone) {
        due = cronMatches(task.cron, toTZ(now, task.timezone));
      }
      if (!due) continue;
      // Recurring cron tasks re-fire on every matching minute; one-shot
      // tasks are marked so they dispatch exactly once.
      if (task.scheduleKind !== "cron") this.dueMarked.add(task.id);
      try {
        await this.services.activity.emit({
          projectId: task.projectId,
          type: "task",
          actor: "scheduler",
          message: `dispatching task "${task.name}" to agent`,
          taskId: task.id,
        });
        const session = await this.services.sessions.request({
          projectId: task.projectId,
          agentId: task.assigneeAgentId,
          taskId: task.id,
        });
        void this.services.sessions.start(session.id).catch((e) => {
          void this.services.activity.emit({
            projectId: task.projectId,
            type: "task",
            actor: "scheduler",
            message: `session start failed: ${e instanceof Error ? e.message : e}`,
            taskId: task.id,
          });
        });
      } catch (e) {
        this.dueMarked.delete(task.id);
        void this.services.activity.emit({
          projectId: task.projectId,
          type: "task",
          actor: "scheduler",
          message: `dispatch failed: ${e instanceof Error ? e.message : e}`,
          taskId: task.id,
        });
      }
    }
  }

  private async fireAutomations(): Promise<void> {
    const all = await this.db.select().from(automations).all();
    const now = this.clock.now();
    // Per-automation last-fire tracking: keep in memory for the process
    // lifetime; a fresh boot may re-fire within the same minute, which is
    // acceptable for a single-operator system (documented).
    for (const a of all) {
      if (a.cron && cronMatches(a.cron, toTZ(now, a.timezone))) {
        const last = this.lastFires.get(a.id);
        const key = nowKey(now);
        if (last === key) continue;
        this.lastFires.set(a.id, key);
        try {
          await this.fireAutomation(a.id);
        } catch (e) {
          void this.services.activity.emit({
            projectId: a.projectId,
            type: "automation",
            actor: "scheduler",
            message: `automation "${a.name}" failed: ${e instanceof Error ? e.message : e}`,
          });
        }
      }
    }
  }

  private lastFires = new Map<string, string>();

  private async fireAutomation(automationId: string): Promise<void> {
    const a = await this.db.select().from(automations).where(eq(automations.id, automationId)).get();
    if (!a) return;
    const agent = await this.db.select().from(agents).where(eq(agents.id, a.agentId)).get();
    if (!agent) throw new Error(`automation agent missing: ${a.agentId}`);

    let task: Task;
    if (a.taskTemplateId) {
      const tpl = await this.db.select().from(taskTemplates).where(eq(taskTemplates.id, a.taskTemplateId)).get();
      if (!tpl) throw new Error(`automation template missing: ${a.taskTemplateId}`);
      const chain = await this.services.tasks.instantiateTemplate(a.projectId, tpl.id, {
        branchName: `auto-${a.id.slice(0, 8)}`,
        featureTitle: a.name,
      });
      task = chain[0]!;
    } else {
      task = await this.services.tasks.create(a.projectId, {
        name: a.name,
        description: a.taskBody ?? `Automation "${a.name}" fired at ${this.clock.now().toISOString()}`,
        assigneeAgentId: agent.id,
        scheduleKind: "now",
      });
    }
    await this.services.activity.emit({
      projectId: a.projectId,
      type: "automation",
      actor: "scheduler",
      message: `automation "${a.name}" fired → task "${task.name}"`,
      taskId: task.id,
    });
  }

  /** Re-scan a single task immediately (used after human marks gated step done). */
  async pokeTask(taskId: string): Promise<void> {
    this.dueMarked.delete(taskId);
    await this.tick();
  }
}

// ---------------------------------------------------------------------------
// Cron
// ---------------------------------------------------------------------------

function toTZ(d: Date, tz: string): Date {
  // SQLite has no TZ tables; approximate with the offset if the tz is a
  // fixed-offset form, else treat as UTC. Labeled approximation.
  const m = /^UTC([+-]\d{1,2})(?::?(\d{2}))?$/.exec(tz);
  if (m) {
    const off = Number(m[1]) * 60 + Number(m[2] ?? 0) * (Number(m[1]) < 0 ? -1 : 1);
    return new Date(d.getTime() + off * 60_000);
  }
  return d;
}

function nowKey(d: Date): string {
  return `${d.getUTCFullYear()}-${d.getUTCMonth()}-${d.getUTCDate()}-${d.getUTCHours()}-${d.getUTCMinutes()}`;
}

/** 5-field cron matcher. */
export function cronMatches(expr: string, date: Date): boolean {
  const fields = expr.trim().split(/\s+/);
  if (fields.length !== 5) throw new Error(`invalid cron: ${expr}`);
  const [minute, hour, dom, month, dow] = fields.map((f, i) => parseField(f, i));
  const m = date.getUTCMinutes();
  const h = date.getUTCHours();
  const d = date.getUTCDate();
  const mo = date.getUTCMonth() + 1;
  const w = date.getUTCDay();
  return minute(m) && hour(h) && dom(d) && month(mo) && dow(w);
}

function parseField(field: string, index: number): (v: number) => boolean {
  const range: [number, number] =
    index === 0 ? [0, 59] : index === 1 ? [0, 23] : index === 2 ? [1, 31] : index === 3 ? [1, 12] : [0, 6];
  const [lo, hi] = range;
  const values = new Set<number>();
  for (const part of field.split(",")) {
    let step = 1;
    let body = part;
    const stepIdx = part.indexOf("/");
    if (stepIdx >= 0) {
      step = Number(part.slice(stepIdx + 1));
      body = part.slice(0, stepIdx);
    }
    const add = (from: number, to: number) => {
      for (let v = from; v <= to; v += step) values.add(v);
    };
    if (body === "*") add(lo, hi);
    else if (body.includes("-")) {
      const [a, b] = body.split("-").map(Number);
      add(Math.max(a, lo), Math.min(b, hi));
    } else {
      values.add(Number(body));
    }
  }
  return (v: number) => values.has(v);
}

File: agentos/services/scheduler.py
from __future__ import annotations
"""
Scheduler: fires scheduled/recurring tasks and cron automations. Port of src/services/scheduler.ts.
Uses APScheduler for the tick loop instead of setInterval.
"""
import asyncio
from datetime import datetime, timezone

from sqlalchemy import select

from agentos.db.client import get_session
from agentos.db.models import Agent as AgentRow, Automation as AutomationRow, Task as TaskRow, TaskTemplate as TaskTemplateRow


def _cron_matches(expr: str, date: datetime) -> bool:
    """5-field cron matcher (minute hour dom month dow)."""
    fields = expr.strip().split()
    if len(fields) != 5:
        return False

    def parse_field(field: str, lo: int, hi: int):
        vals: set[int] = set()
        for part in field.split(","):
            step = 1
            if "/" in part:
                body, step_s = part.rsplit("/", 1)
                step = int(step_s)
            else:
                body = part
            if body == "*":
                for v in range(lo, hi + 1, step):
                    vals.add(v)
            elif "-" in body:
                a, b = body.split("-")
                for v in range(int(a), int(b) + 1, step):
                    vals.add(v)
            else:
                vals.add(int(body))
        return vals

    minute, hour, dom, month, dow = fields
    return (
        date.minute in parse_field(minute, 0, 59)
        and date.hour in parse_field(hour, 0, 23)
        and date.day in parse_field(dom, 1, 31)
        and date.month in parse_field(month, 1, 12)
        and date.weekday() in parse_field(dow, 0, 6)  # 0=Mon in Python; cron 0=Sun — approximate
    )


class SchedulerService:
    def __init__(self, services: object, interval_ms: int = 1000) -> None:
        self._svc = services
        self._interval = interval_ms / 1000
        self._cloud_busy = False
        self._ticking = False
        self._due_marked: set[str] = set()
        self._last_fires: dict[str, str] = {}
        self._task: asyncio.Task | None = None

    def start(self) -> None:
        if self._task and not self._task.done():
            return
        self._task = asyncio.ensure_future(self._loop())

    def stop(self) -> None:
        if self._task:
            self._task.cancel()
            self._task = None

    def is_cloud_busy(self) -> bool:
        return self._cloud_busy

    def set_cloud_busy(self, busy: bool) -> None:
        self._cloud_busy = busy

    async def _loop(self) -> None:
        while True:
            await asyncio.sleep(self._interval)
            try:
                await self.tick()
            except Exception as e:
                try:
                    await self._svc.activity.emit(  # type: ignore[attr-defined]
                        type="system", actor="scheduler", message=f"tick error: {e}"
                    )
                except Exception:
                    pass

    async def tick(self) -> None:
        if self._ticking:
            return
        self._ticking = True
        try:
            await self._dispatch_due_tasks()
            await self._fire_automations()
        finally:
            self._ticking = False

    async def _dispatch_due_tasks(self) -> None:
        now = datetime.now(timezone.utc)
        async with get_session() as db:
            result = await db.execute(select(TaskRow))
            all_tasks = result.scalars().all()

        for row in all_tasks:
            # Recurring cron: reset done → todo
            if row.status == "done" and row.schedule_kind == "cron":
                async with get_session() as db:
                    result = await db.execute(select(TaskRow).where(TaskRow.id == row.id))
                    t = result.scalar_one_or_none()
                    if t:
                        activity = list(t.activity or [])
                        activity.append({"at": now.isoformat(), "actor": "scheduler",
                                         "message": "recurring task reset for next run"})
                        t.status = "todo"
                        t.activity = activity
                        await db.commit()
                row.status = "todo"

            if row.status != "todo":
                self._due_marked.discard(row.id)
                continue
            if row.assignee_type != "agent" or not row.assignee_agent_id:
                continue
            if row.id in self._due_marked:
                continue
            if self._svc.sessions.is_task_running(row.id):  # type: ignore[attr-defined]
                continue

            # Chain blocking
            if row.chain_id and row.chain_index and row.chain_index > 0:
                prev = next(
                    (t for t in all_tasks
                     if t.chain_id == row.chain_id and t.chain_index == row.chain_index - 1),
                    None,
                )
                if not prev or prev.status != "done":
                    continue

            due = False
            if row.schedule_kind == "now":
                due = True
            elif row.schedule_kind == "at" and row.run_at:
                due = now >= datetime.fromisoformat(row.run_at)
            elif row.schedule_kind == "cron" and row.cron:
                due = _cron_matches(row.cron, now)

            if not due:
                continue
            if row.schedule_kind != "cron":
                self._due_marked.add(row.id)

            try:
                await self._svc.activity.emit(  # type: ignore[attr-defined]
                    project_id=row.project_id, type="task", actor="scheduler",
                    message=f'dispatching task "{row.name}" to agent', task_id=row.id,
                )
                session = await self._svc.sessions.request(  # type: ignore[attr-defined]
                    project_id=row.project_id, agent_id=row.assignee_agent_id, task_id=row.id
                )
                asyncio.ensure_future(self._start_session(session["id"], row.project_id, row.id, row.name))
            except Exception as e:
                self._due_marked.discard(row.id)
                await self._svc.activity.emit(  # type: ignore[attr-defined]
                    project_id=row.project_id, type="task", actor="scheduler",
                    message=f"dispatch failed: {e}", task_id=row.id,
                )

    async def _start_session(self, session_id: str, project_id: str, task_id: str, task_name: str) -> None:
        try:
            await self._svc.sessions.start(session_id)  # type: ignore[attr-defined]
        except Exception as e:
            await self._svc.activity.emit(  # type: ignore[attr-defined]
                project_id=project_id, type="task", actor="scheduler",
                message=f"session start failed: {e}", task_id=task_id,
            )

    async def _fire_automations(self) -> None:
        now = datetime.now(timezone.utc)
        async with get_session() as db:
            result = await db.execute(select(AutomationRow))
            autos = result.scalars().all()

        for a in autos:
            if not a.cron or not _cron_matches(a.cron, now):
                continue
            key = f"{now.year}-{now.month}-{now.day}-{now.hour}-{now.minute}"
            if self._last_fires.get(a.id) == key:
                continue
            self._last_fires[a.id] = key
            try:
                await self._fire_automation(a.id)
            except Exception as e:
                await self._svc.activity.emit(  # type: ignore[attr-defined]
                    project_id=a.project_id, type="automation", actor="scheduler",
                    message=f'automation "{a.name}" failed: {e}',
                )

    async def _fire_automation(self, automation_id: str) -> None:
        async with get_session() as db:
            result = await db.execute(select(AutomationRow).where(AutomationRow.id == automation_id))
            a = result.scalar_one_or_none()
            if not a:
                return
            result2 = await db.execute(select(AgentRow).where(AgentRow.id == a.agent_id))
            agent = result2.scalar_one_or_none()
            if not agent:
                raise ValueError(f"automation agent missing: {a.agent_id}")

        now = datetime.now(timezone.utc).isoformat()
        if a.task_template_id:
            chain = await self._svc.tasks.instantiate_template(  # type: ignore[attr-defined]
                a.project_id, a.task_template_id,
                {"branchName": f"auto-{a.id[:8]}", "featureTitle": a.name},
            )
            task = chain[0]
        else:
            task = await self._svc.tasks.create(a.project_id, {  # type: ignore[attr-defined]
                "name": a.name,
                "description": a.task_body or f'Automation "{a.name}" fired at {now}',
                "assigneeAgentId": agent.id,
                "scheduleKind": "now",
            })
        await self._svc.activity.emit(  # type: ignore[attr-defined]
            project_id=a.project_id, type="automation", actor="scheduler",
            message=f'automation "{a.name}" fired → task "{task["name"]}"', task_id=task["id"],
        )

    async def poke_task(self, task_id: str) -> None:
        self._due_marked.discard(task_id)
        await self.tick()
